### PR TITLE
A3: Number Conversion Game

### PR DESCRIPTION
**Pull request by sudarson-25**

### CODE REVIEW

**Comment by RaaneshKumar** [comment]
Handle cases for non-decimal inputs.
image

File: Training/Training/main.c [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+         hexNeg[k] = hexNegRev[mCount - k - 1];
+      hexNeg[k] = '\0';
+      free (hexNegRev);
+      return hexNeg;
+   }
+}
+
+int main () {
+   int dec;
+   printf ("Input (Enter an integer): ");
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Ask user inputs in an infinite loop such that we can check the program multiple times in a single run. In your case the program ends for each try.

File: Training/Training/main.c [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+      rem = dec % 2;
+      dec /= 2;
+      mCount++;
+   }
+   // Representing binary in multiples of 8-bits accordingly
+   if (mCount % 8 != 0)
+      mCount += (8 - (mCount % 8));
+   int* binRev = NULL;
+   do {
+      binRev = (int*)malloc (mCount * sizeof (int));
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Can't we just have an array for this process?

File: Training/Training/main.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+   int dec2 = dec, rem;
+   if (dec < 0)
+      dec *= -1;
+   int dec1 = dec;
+   //Counting the number of digits in binary representation
+   while (dec > 0) {
+      rem = dec % 2;
+      dec /= 2;
+      mCount++;
+   }
```
Comment (RaaneshKumar) [inline Training/Training/main.c:33]: This part of code is getting repeated multiple times which I think is unnecessary. You can collect all the remainders in a collection and use them further wherever required and the count of the collection will itself do the job of mCount. Try this. Even if this part of code is required for sure in multiple parts of the code move it to a method and call it wherever necessary to avoid repetition of the code.

File: Training/Training/main.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      }
+      free (binRev);
+      //Adding 1 to the the binary number (stored as individual digits)
+      for (int k = mCount - 1; k >= 0; k--)
+         if (binTwosComp[k] == 0) {
+            binTwosComp[k] = 1;
+            break;
+         } else {
+            binTwosComp[k] = 0;
+         }
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Brackets can be removed for single line statements in if-else conditions.

**Review by RaaneshKumar — CHANGES_REQUESTED** [review]

**Comment by RaaneshKumar** [comment]
Unassign yourself before assigning others. There should be only one assignee at a time.

File: Training/Training/main.c [hunk 5]
```diff
@@ -0,0 +1,185 @@
+#include<stdio.h>
+#include<malloc.h>
+#include<math.h>
+
+int mCount;
+
+//Function for counting the number of digits
+int count (int dec, int div) {
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Method name must begin with an uppercase letter.

File: Training/Training/main.c [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+#include<malloc.h>
+#include<math.h>
+
+int mCount;
+
+//Function for counting the number of digits
+int count (int dec, int div) {
+   int c = 0;
+   if (dec == 0)
+      return 8;
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Can be written in the same line. Apply wherever possible.

File: Training/Training/main.c [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+      } while (hexNegRev == NULL);
+      int ind = 0;
+      //Converting binary to hex and storing into hexNegRev in reverse
+      for (int i = mCount - 1; i >= 0; i--) {
+         digit += bin[i] * pow (2, place);
+         place++;
+         icount++;
+         if (icount == 4) {
+            place = 0;
+            icount = 0;
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Can be written in a single line.
place = 0, icount = 0;

File: Training/Training/main.c [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+      return binTwosComp;
+   }
+}
+
+char* DecToHex (int dec) {
+   mCount = 0;
+   int dec2 = dec;
+   if (dec < 0)
+      dec *= -1;
+   int dec1 = dec;
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: The names for variables dec1 and dec2 can be more meaningful. For example, int dec, dec2, dec1 can be named as n, nTemp, negN respectively which will explain its purpose by just reading.

File: Training/Training/main.c [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+   do {
+      hexRev = (char*)malloc (mCount * sizeof (char));
+   } while (hexRev == NULL);
+   //Storing the remainders into hexRev
+   for (int i = 0; i < mCount; i++) {
+      if (dec1 % 16 >= 10)
+         hexRev[i] = (dec1 % 16) + 55;
+      else
+         hexRev[i] = (dec1 % 16) + 48;
+      dec1 /= 16;
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Can be written using ternary operators. Will look like this.
hexRev [i] = dec1 % 16 >= 10 ? (dec1 % 16) + 55 : (dec1 % 16) + 48;
Apply wherever possible.

File: Training/Training/main.c [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+         hexNeg[k] = hexNegRev[mCount - k - 1];
+      hexNeg[k] = '\0';
+      free (hexNegRev);
+      return hexNeg;
+   }
+}
+
+int main () {
+   int dec;
+   char term;
```
Comment (RaaneshKumar) [inline Training/Training/main.c:153]: dec = 0 can be handled in main itself. Otherwise, it will have to run through the whole program to just print the very known output 0.

**Review by RaaneshKumar — CHANGES_REQUESTED** [review]

**Comment by RaaneshKumar** [comment]
Make sure your commits are in the form "A3: Your message.". This is the format for commits used in Metamation. Stick to it hereafter. Also change the title from "Assignment3" to "A3: Number Conversion Game". Follow this format for titles in the upcoming assignments also.

File: Training/Training/main.c [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+      return bin;
+   } else { //Performing two's compliment
+      int* binTwosComp = NULL;
+      do {
+         binTwosComp = (int*)malloc (mCount * sizeof (int));
+      } while (binTwosComp == NULL);
+      //Complimenting each element of the array
+      for (int k = 0; k < mCount; k++) {
+         binTwosComp[k] = binRev[mCount - k - 1] == 0 ? 1 : 0;
+      }
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Brackets can be removed.

File: Training/Training/main.c [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+      int ind = 0;
+      //Converting binary to hex and storing into hexNegRev in reverse
+      for (int i = mCount - 1; i >= 0; i--) {
+         digit += bin[i] * pow (2, place);
+         place++;
+         icount++;
+         if (icount == 4) {
+            place = 0, icount = 0;
+            hexNegRev[ind] = digit >= 10 ? digit + 55 : digit + 48;
+            ind++;
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Can be written in a single line. Will look like
hexNegRev[ind++] = digit >= 10 ? digit + 55 : digit + 48;
Make this change wherever applicable.

File: Training/Training/main.c [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+            term = fgetc (stdin);
+            if (term == '\n') break;
+         }
+         printf ("\n\n");
+      } else {
+         if (dec == 0) {
+            int converted[8];
+            for (int i = 0; i < 8; i++) {
+               converted[i] = 0;
+            }
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Brackets can be removed for single line statements. Make this change wherever applicable. Review your code once before pushing.

**Review by RaaneshKumar — CHANGES_REQUESTED** [review]

**Comment by RaaneshKumar** [comment]
Don't make changes and commit when it is assigned to someone else. If you need to make a change, please assign it to yourself and then do your change.

File: Training/Training/main.c [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+               converted[i] = 0;
+            printf ("Binary (%d-bit): ", 8);
+            for (int i = 0; i < 8; i++) {
+               printf ("%d", converted[i]);
+            }
+            printf ("\nHEX (%d-bit): ", 8);
+            for (int i = 0; i < 8; i++) {
+               printf ("%d", converted[i]);
+            }
+            printf ("\n\n");
```
Comment (RaaneshKumar) [inline Training/Training/main.c]: Not necessary. If dec = 0, you can simply print the answer as a string itself.

**Review by RaaneshKumar — CHANGES_REQUESTED** [review]

File: Training/Training/Training.vcxproj [hunk 15]
```diff
@@ -127,7 +127,7 @@
     </Link>
   </ItemDefinitionGroup>
   <ItemGroup>
-    <ClCompile Include="Training.c" />
```
Comment (Dakshita-Minocha) [inline Training/Training/Training.vcxproj:130]: You don't need to rename the file everytime. Trainng.c is okay, unless you add more files and need to give context.

File: Training/Training/main.c [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+// ------------------------------------------------------------------------------------------------
+
+#include<stdio.h>
+#include<malloc.h>
+#include<math.h>
+
+int mCount;
+
+//Function for counting the number of digits
+int Count (int dec, int div) {
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c]: DigitCount may be a more appropriate name.

File: Training/Training/main.c [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+   }
+   return c;
+}
+
+int* DecToBin (int dec) {
+   mCount = 0;
+   int decReal = dec;
+   if (dec < 0) dec *= -1;
+   int decTemp = dec;
+   //Counting the number of digits in binary representation
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c]: You do not need to add comments like this if you document (comment) your functions properly

File: Training/Training/main.c [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+// ------------------------------------------------------------------
+// main.c
+// Program to convert and display decimal value into binary and hexadecimal value.
+// ------------------------------------------------------------------------------------------------
+
+#include<stdio.h>
+#include<malloc.h>
+#include<math.h>
+
+int mCount;
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c]: If the only reason you have declared mCount as global is that you don't want to declare it again, it is not required. You should declare it within the functions. Otherwise please let me know the reason.

File: Training/Training/main.c [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+
+//Function for counting the number of digits
+int Count (int dec, int div) {
+   int c = 0;
+   while (dec > 0) {
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c]: Count calculates log2 here. You can use log2 () from Math.h. Similarly use log10 where required.

File: Training/Training/main.c [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+   if (decReal > 0) {
+      int* bin;
+      do {
+         bin = (int*)malloc (mCount * sizeof (int));
+      } while (bin == NULL);
+      // Reverse storing the stored remainders
+      for (int j = 0; j < mCount; j++)
+         bin[j] = binRev[mCount - j - 1];
+      return bin;
+   } else { //Performing two's compliment
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c:53]: Also, it's nice that you've tried to implement 2's complement logic manually, but there is an easier way to do this. Explore bit shift operators as an alternative.

**Review by Dakshita-Minocha — COMMENTED** [review]
Please write a test with edge cases and show if your code passes / fails these tests.

File: Training/Training/main.c [hunk 21]
```diff
@@ -4,47 +4,51 @@
...
+#include<string.h>
+
+#define BYTESIZE 8
+#define ANSI_COLOR_RED     "\x1b[31m"
+#define ANSI_COLOR_GREEN   "\x1b[32m"
+#define ANSI_COLOR_RESET   "\x1b[0m"
+#define ANSI_COLOR_YELLOW  "\x1b[33m"
 
 /// <summary>
 /// Function that counts the number of bits the binary/hex representation will occupy
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c:23]:
```suggestion
/// Returns number of bits required to represent number in multiples of 8.
```

File: Training/Training/main.c [hunk 22]
```diff
@@ -4,47 +4,51 @@
...
+/// and alters the count to the succeeding multiple of 8.
 /// </summary>
-int DigitsCount (long long int dec, int div) {
-   return (int)((log ((double)dec) / log (div)) + 1);
+int DigitsCount (int dec, int div) {
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c:26]: dev and div don't indicate what they are. num may be a better alternative (for dec), base for div

File: Training/Training/main.c [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+   int userinputs[] = { 0,INT_MAX,-4321,1234,5678 };
+   length = sizeof (userinputs) / sizeof (userinputs[0]);
+   char* binexpected[] = { "00000000", "01111111111111111111111111111111", "1110111100011111", "0000010011010010", "0001011000101110" };
+   char* hexexpected[] = { "00000000","7FFFFFFF","FFFFEF1F", "000004D2", "0000162E" };
+   TestCases (userinputs, length, binexpected, hexexpected);
+   printf ("Data Convertor\n");
+   printf ("~~~~~~~~~~~~~~\n");
+   while (1) {
+      printf ("Input (Enter an integer): ");
+      if (scanf_s ("%d%c", &dec, &term, 1) != 2 || term != '\n') {
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c:163]: [scanf should be preferred for formatted input](https://c-faq.com/stdio/scanfprobs.html). For inputs from stdin, it is better to use fgets or other methods that read the whole line at once.

File: Training/Training/main.c [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+         printf ("Enter a valid input!\n\n");
+         //Draining the unassigned inputs from stdin
+         for (; ; ) {
+            term = fgetc (stdin);
+            if (term == '\n') break;
+         }
+      } else if (dec < INT_MIN || dec > INT_MAX)
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c:170]: This line of code is never hit 
image

File: Training/Training/main.c [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+   }
+   printf ("Decimal to Hex\n");
+   printf ("~~~~~~~~~~~~~~\n");
+   for (int i = 0; i < length; i++) {
+      printf (ANSI_COLOR_YELLOW"Case %d"ANSI_COLOR_RESET"\n", i + 1);
+      printf ("Input: %d\n", userinputs[i]);
+      char* hex = DecToHex (userinputs[i], &digits);
+      if (strcmp (hex, hexexpected[i]) == 0)
+         printf (ANSI_COLOR_GREEN"Passed"ANSI_COLOR_RESET"\n\n");
+      else
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c:146]:
```suggestion
      if else
printf (flag == digits ? ANSI_COLOR_GREEN"Passed"ANSI_COLOR_RESET"\n\n" : ANSI_COLOR_RED"Failed"ANSI_COLOR_RESET"\n\n");
```

File: Training/Training/main.c [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+   printf ("Decimal to Binary\n");
+   printf ("~~~~~~~~~~~~~~~~~\n");
+   for (int i = 0; i < length; i++) {
+      printf (ANSI_COLOR_YELLOW"Case %d"ANSI_COLOR_RESET"\n", i + 1);
+      printf ("Input: %d\n", userinputs[i]);
+      int* bin = DecToBin (userinputs[i], &digits);
+      flag = 0;
+      for (int j = 0; j < digits; j++) {
+         temp = bin[j] + 48;
+         if (temp == binexpected[i][j])
```
Comment (Dakshita-Minocha) [inline Training/Training/main.c:130]:
```suggestion
         if (temp == binexpected[i][j]) flag++;
```

**Review by Dakshita-Minocha — APPROVED** [review]
